stdlib/public/core/CollectionDifference.swift

```diff
@@ -0,0 +1,378 @@
+      switch (a, b) {
+      case (.remove(_, _, _), .insert(_, _, _)):
+        return true
+      case (.insert(_, _, _), .remove(_, _, _)):
+        return false
+      default:
+        return a._offset < b._offset
+      }
+    }
+
+    // Find first insertion via binary search
+    let firstInsertIndex: Int
+    if sortedChanges.count == 0 {
+      firstInsertIndex = 0
+    } else {
+      var range = 0...sortedChanges.count
+      while range.lowerBound != range.upperBound {
+        let i = (range.lowerBound + range.upperBound) / 2
+        switch sortedChanges[i] {
+        case .insert(_, _, _):
+          range = range.lowerBound...i
+        case .remove(_, _, _):
+          range = (i + 1)...range.upperBound
+        }
+      }
+      firstInsertIndex = range.lowerBound
+    }
+
+    removals = Array(sortedChanges[0..<firstInsertIndex])
+    insertions = Array(sortedChanges[firstInsertIndex..<sortedChanges.count])
+  }
+}
+
+/// A CollectionDifference is itself a Collection.
+///
+/// The enumeration order of `Change` elements is:
+///
+/// 1. `.remove`s, from highest `offset` to lowest
+/// 2. `.insert`s, from lowest `offset` to highest
+///
+/// This guarantees that applicators on compatible base states are safe when
+/// written in the form:
+///
+/// ```
+/// for c in diff {
+///   switch c {
+///   case .remove(offset: let o, element: _, associatedWith: _):
+///     arr.remove(at: o)
+///   case .insert(offset: let o, element: let e, associatedWith: _):
+///     arr.insert(e, at: o)
+///   }
+/// }
+/// ```
+@available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *) // FIXME(availability-5.1)
+extension CollectionDifference: Collection {
+  public typealias Element = Change
+
+  public struct Index: Equatable, Hashable, Comparable {
+    // Opaque index type is isomorphic to Int
+    internal let _offset: Int
+
+    internal init(_offset offset: Int) {
+      _offset = offset
+    }
+  }
+
+  public var startIndex: Index {
+    return Index(_offset: 0)
+  }
+
+  public var endIndex: Index {
+    return Index(_offset: removals.count + insertions.count)
+  }
+
+  public func index(after index: Index) -> Index {
+    return Index(_offset: index._offset + 1)
+  }
+
+  public subscript(position: Index) -> Element {
+    if position._offset < removals.count {
+      return removals[removals.count - (position._offset + 1)]
+    }
+    return insertions[position._offset - removals.count]
+  }
+
+  public func index(before index: Index) -> Index {
+    return Index(_offset: index._offset - 1)
+  }
+
+  public func formIndex(_ index: inout Index, offsetBy distance: Int) {
+    index = Index(_offset: index._offset + distance)
+  }
+
+  public func distance(from start: Index, to end: Index) -> Int {
+    return end._offset - start._offset
+  }
+}
+
+@available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *) // FIXME(availability-5.1)
+extension CollectionDifference.Index { // Comparable
```

Comment:
Can't the conformance be removed from the declaration above, and uncommented here?